File: iosApp/UkuleleCompanion/ViewModels/SongbookViewModel.swift
import Foundation
import shared

struct StoredSong: Codable, Identifiable {
    let id: String
    var title: String
    var subtitle: String
    var artist: String
    var content: String
    var key: String
    var capo: Int
    var strumPatternName: String
    var labels: [String]
    var createdAt: Double
    var updatedAt: Double
    var viewCount: Int
    var lastViewedAt: Double
    var totalViewTimeMs: Double

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(String.self, forKey: .id)
        title = try container.decode(String.self, forKey: .title)
        subtitle = try container.decodeIfPresent(String.self, forKey: .subtitle) ?? ""
        artist = try container.decode(String.self, forKey: .artist)
        content = try container.decode(String.self, forKey: .content)
        key = try container.decode(String.self, forKey: .key)
        capo = try container.decode(Int.self, forKey: .capo)
        strumPatternName = try container.decode(String.self, forKey: .strumPatternName)
        labels = try container.decode([String].self, forKey: .labels)
        createdAt = try container.decode(Double.self, forKey: .createdAt)
        updatedAt = try container.decode(Double.self, forKey: .updatedAt)
        viewCount = try container.decodeIfPresent(Int.self, forKey: .viewCount) ?? 0
        lastViewedAt = try container.decodeIfPresent(Double.self, forKey: .lastViewedAt) ?? 0
        totalViewTimeMs = try container.decodeIfPresent(Double.self, forKey: .totalViewTimeMs) ?? 0
    }

    init(id: String, title: String, subtitle: String = "", artist: String, content: String,
         key: String, capo: Int, strumPatternName: String, labels: [String],
         createdAt: Double, updatedAt: Double,
         viewCount: Int = 0, lastViewedAt: Double = 0, totalViewTimeMs: Double = 0) {
        self.id = id
        self.title = title
        self.subtitle = subtitle
        self.artist = artist
        self.content = content
        self.key = key
        self.capo = capo
        self.strumPatternName = strumPatternName
        self.labels = labels
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.viewCount = viewCount
        self.lastViewedAt = lastViewedAt
        self.totalViewTimeMs = totalViewTimeMs
    }
}

enum SongSortOrder: String, CaseIterable {
    case lastModified = "Last Modified"
    case dateAdded = "Date Added"
    case title = "Title"
    case artist = "Artist"
}

@MainActor
final class SongbookViewModel: ObservableObject {
    @Published var songs: [StoredSong] = []
    @Published var searchQuery = ""
    @Published var sortOrder: SongSortOrder = .lastModified
    @Published var selectedLabels: Set<String> = []

    private let repository: SongbookRepository

    init(repository: SongbookRepository = SongbookRepository()) {
        self.repository = repository
        songs = repository.getAll()
    }

    var allLabels: [String] {
        let labels = songs.flatMap { $0.labels }
        return Array(Set(labels)).sorted()
    }

    var filteredSongs: [StoredSong] {
        var result = songs

        if !searchQuery.isEmpty {
            let q = searchQuery.lowercased()
            result = result.filter {
                $0.title.lowercased().contains(q) ||
                $0.artist.lowercased().contains(q)
            }
        }

        if !selectedLabels.isEmpty {
            result = result.filter { song in
                selectedLabels.isSubset(of: Set(song.labels))
            }
        }

        switch sortOrder {
        case .lastModified:
            result.sort { $0.updatedAt > $1.updatedAt }
        case .dateAdded:
            result.sort { $0.createdAt > $1.createdAt }
        case .title:
            result.sort { $0.title.localizedCaseInsensitiveCompare($1.title) == .orderedAscending }
        case .artist:
            result.sort { $0.artist.localizedCaseInsensitiveCompare($1.artist) == .orderedAscending }
        }

        return result
    }

    func save(song: StoredSong) {
        if let idx = songs.firstIndex(where: { $0.id == song.id }) {
            var updated = song
            updated.updatedAt = Date().timeIntervalSince1970 * 1000
            songs[idx] = updated
        } else {
            songs.insert(song, at: 0)
        }
        repository.save(songs)
    }

    func duplicate(song: StoredSong) -> StoredSong {
        let now = Date().timeIntervalSince1970 * 1000
        let copy = StoredSong(
            id: UUID().uuidString,
            title: "\(song.title) (Copy)",
            subtitle: song.subtitle,
            artist: song.artist,
            content: song.content,
            key: song.key,
            capo: song.capo,
            strumPatternName: song.strumPatternName,
            labels: song.labels,
            createdAt: now,
            updatedAt: now
        )
        songs.insert(copy, at: 0)
        repository.save(songs)
        return copy
    }

    func delete(id: String) {
        songs.removeAll { $0.id == id }
        repository.save(songs)
    }

    func updateLabels(id: String, labels: [String]) {
        guard let idx = songs.firstIndex(where: { $0.id == id }) else { return }
        songs[idx].labels = labels
        songs[idx].updatedAt = Date().timeIntervalSince1970 * 1000
        repository.save(songs)
    }

    func updateStrumPattern(id: String, patternName: String) {
        guard let idx = songs.firstIndex(where: { $0.id == id }) else { return }
        songs[idx].strumPatternName = patternName
        songs[idx].updatedAt = Date().timeIntervalSince1970 * 1000
        repository.save(songs)
    }

    func recordView(id: String) {
        guard let idx = songs.firstIndex(where: { $0.id == id }) else { return }
        songs[idx].viewCount += 1
        songs[idx].lastViewedAt = Date().timeIntervalSince1970 * 1000
        repository.save(songs)
    }

    func recordViewTime(id: String, elapsedMs: Double) {
        guard elapsedMs > 0,
              let idx = songs.firstIndex(where: { $0.id == id }) else { return }
        songs[idx].totalViewTimeMs += elapsedMs
        repository.save(songs)
    }

    /// Derives a song title from an imported file name.
    ///
    /// Strips any path and document-ID prefix before dropping the extension, so
    /// a URI or document ID can never surface as the title (issue #500).
    ///
    /// - Parameter filename: The resolved file name, or `nil` when unavailable.
    /// - Returns: A display title, falling back to `"Imported Song"`.
    nonisolated static func titleFromFilename(_ filename: String?) -> String {
        guard let filename else { return defaultImportTitle }
        let base = filename
            .components(separatedBy: "/").last?
            .components(separatedBy: ":").last?
            .replacingOccurrences(of: "_", with: " ")
            .trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        let withoutExtension = (base as NSString).deletingPathExtension
            .trimmingCharacters(in: .whitespacesAndNewlines)
        return withoutExtension.isEmpty ? defaultImportTitle : withoutExtension
    }

    private nonisolated static let defaultImportTitle = "Imported Song"

    func importPlainText(content: String, filename: String?) {
        let now = Date().timeIntervalSince1970 * 1000
        let song = StoredSong(
            id: UUID().uuidString,
            title: Self.titleFromFilename(filename),
            artist: "",
            content: content.trimmingCharacters(in: .whitespacesAndNewlines),
            key: "",
            capo: 0,
            strumPatternName: "",
            labels: [],
            createdAt: now,
            updatedAt: now
        )
        save(song: song)
    }

    /// Imports a song from ChordPro text.
    ///
    /// - Parameters:
    ///   - text: The raw ChordPro content.
    ///   - filename: Optional file name used as a fallback title when the
    ///     content carries no `{title}` directive.
    func importChordPro(text: String, filename: String? = nil) {
        let sheet = ChordProParser.shared.parse(
            input: text,
            defaultTitle: Self.titleFromFilename(filename)
        )
        let song = StoredSong(
            id: sheet.id,
            title: sheet.title,
            subtitle: sheet.subtitle,
            artist: sheet.artist,
            content: sheet.content,
            key: sheet.key,
            capo: Int(sheet.capo),
            strumPatternName: sheet.strumPatternName,
            labels: sheet.labels.asStrings,
            createdAt: Double(sheet.createdAt),
            updatedAt: Double(sheet.updatedAt)
        )
        save(song: song)
    }

    func transpose(song: StoredSong, semitones: Int) -> StoredSong {
        let transposed = ChordSheetTranspose.shared.transpose(
            content: song.content,
            semitones: Int32(semitones)
        )
        var updated = song
        updated.content = transposed
        updated.updatedAt = Date().timeIntervalSince1970 * 1000
        return updated
    }

    func formattedDisplay(song: StoredSong) -> String {
        let sheet = toChordSheet(song)
        return ChordSheetFormatter.shared.formatChordsAboveLyrics(sheet: sheet)
    }

    func exportChordPro(song: StoredSong) -> String {
        let sheet = toChordSheet(song)
        return ChordProExporter.shared.export(sheet: sheet)
    }

    private func toChordSheet(_ song: StoredSong) -> ChordSheet {
        ChordSheet(
            id: song.id,
            title: song.title,
            subtitle: song.subtitle,
            artist: song.artist,
            content: song.content,
            key: song.key,
            capo: Int32(song.capo),
            strumPatternName: song.strumPatternName,
            labels: song.labels,
            createdAt: Int64(song.createdAt),
            updatedAt: Int64(song.updatedAt),
            viewCount: Int32(song.viewCount),
            lastViewedAt: Int64(song.lastViewedAt),
            totalViewTimeMs: Int64(song.totalViewTimeMs)
        )
    }

    func importData(_ incoming: [[String: Any]]) {
        repository.importData(incoming, into: &songs)
    }

    func exportData() -> [[String: Any]] {
        repository.exportData(songs)
    }
}

File: app/src/main/java/com/baijum/ukufretboard/viewmodel/SongbookViewModel.kt
package com.baijum.ukufretboard.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import com.baijum.ukufretboard.data.ChordProParser
import com.baijum.ukufretboard.data.ChordSheet
import com.baijum.ukufretboard.data.ChordSheetRepository
import com.baijum.ukufretboard.data.SongSortOrder
import com.baijum.ukufretboard.domain.ChordSheetTranspose
import com.baijum.ukufretboard.domain.SongbookFilter
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/**
 * ViewModel for managing the songbook (list of chord sheets).
 */
class SongbookViewModel(application: Application) : AndroidViewModel(application) {

    private val repository = ChordSheetRepository(application)

    private val _allSheets = MutableStateFlow<List<ChordSheet>>(emptyList())

    private val _searchQuery = MutableStateFlow("")
    val searchQuery: StateFlow<String> = _searchQuery.asStateFlow()

    private val _sortOrder = MutableStateFlow(SongSortOrder.LAST_MODIFIED)
    val sortOrder: StateFlow<SongSortOrder> = _sortOrder.asStateFlow()

    private val _selectedLabels = MutableStateFlow<Set<String>>(emptySet())
    val selectedLabels: StateFlow<Set<String>> = _selectedLabels.asStateFlow()

    private val _allLabels = MutableStateFlow<Set<String>>(emptySet())
    /** All distinct labels used across every saved song. */
    val allLabels: StateFlow<Set<String>> = _allLabels.asStateFlow()

    private val _sheets = MutableStateFlow<List<ChordSheet>>(emptyList())

    /** Observable list of filtered and sorted chord sheets. */
    val sheets: StateFlow<List<ChordSheet>> = _sheets.asStateFlow()

    /** The currently open sheet (for viewing/editing). */
    private val _currentSheet = MutableStateFlow<ChordSheet?>(null)
    val currentSheet: StateFlow<ChordSheet?> = _currentSheet.asStateFlow()

    /** Whether the editor is open. */
    private val _isEditing = MutableStateFlow(false)
    val isEditing: StateFlow<Boolean> = _isEditing.asStateFlow()

    init {
        refresh()
    }

    fun refresh() {
        _allSheets.value = repository.getAll()
        _allLabels.value = _allSheets.value.flatMap { it.labels }
            .toSortedSet(String.CASE_INSENSITIVE_ORDER)
        applyFilterAndSort()
    }

    fun setSearchQuery(query: String) {
        _searchQuery.value = query
        applyFilterAndSort()
    }

    fun setSortOrder(order: SongSortOrder) {
        _sortOrder.value = order
        applyFilterAndSort()
    }

    fun toggleLabelFilter(label: String) {
        val current = _selectedLabels.value
        _selectedLabels.value = if (label in current) current - label else current + label
        applyFilterAndSort()
    }

    fun clearLabelFilter() {
        _selectedLabels.value = emptySet()
        applyFilterAndSort()
    }

    private fun applyFilterAndSort() {
        val activeLabels = _selectedLabels.value

        var result = SongbookFilter.apply(
            songs = _allSheets.value,
            query = _searchQuery.value,
            label = activeLabels.firstOrNull(),
            sortOrder = _sortOrder.value,
            title = { it.title },
            artist = { it.artist },
            labels = { it.labels },
            updatedAt = { it.updatedAt },
            createdAt = { it.createdAt },
        )

        if (activeLabels.size > 1) {
            result = result.filter { sheet -> sheet.labels.containsAll(activeLabels) }
        }

        _sheets.value = result
    }

    private var viewOpenedAt: Long = 0L

    fun openSheet(sheet: ChordSheet) {
        val now = System.currentTimeMillis()
        val updated = sheet.copy(
            viewCount = sheet.viewCount + 1,
            lastViewedAt = now,
        )
        repository.save(updated)
        _currentSheet.value = updated
        _isEditing.value = false
        viewOpenedAt = now
        refresh()
    }

    fun closeSheet() {
        recordViewTime()
        _currentSheet.value = null
        _isEditing.value = false
    }

    private fun recordViewTime() {
        val sheet = _currentSheet.value ?: return
        if (viewOpenedAt <= 0L) return
        val elapsed = System.currentTimeMillis() - viewOpenedAt
        viewOpenedAt = 0L
        if (elapsed <= 0) return
        val updated = sheet.copy(totalViewTimeMs = sheet.totalViewTimeMs + elapsed)
        repository.save(updated)
        _currentSheet.value = updated
        refresh()
    }

    fun startEditing(sheet: ChordSheet? = null) {
        _currentSheet.value = sheet ?: ChordSheet(title = "", content = "")
        _isEditing.value = true
    }

    fun saveSheet(
        title: String,
        artist: String,
        content: String,
        key: String = "",
        strumPatternName: String = "",
        labels: List<String> = emptyList(),
    ) {
        val existing = _currentSheet.value
        val sheet = if (existing != null && existing.title.isNotEmpty()) {
            existing.copy(
                title = title,
                artist = artist,
                content = content,
                key = key,
                strumPatternName = strumPatternName,
                labels = labels,
                updatedAt = System.currentTimeMillis(),
            )
        } else {
            ChordSheet(
                title = title,
                artist = artist,
                content = content,
                key = key,
                strumPatternName = strumPatternName,
                labels = labels,
            )
        }
        repository.save(sheet)
        _currentSheet.value = sheet
        _isEditing.value = false
        refresh()
    }

    fun updateLabels(labels: List<String>) {
        val sheet = _currentSheet.value ?: return
        val updated = sheet.copy(
            labels = labels,
            updatedAt = System.currentTimeMillis(),
        )
        repository.save(updated)
        _currentSheet.value = updated
        refresh()
    }

    fun applyTranspose(semitones: Int) {
        if (semitones == 0) return
        val sheet = _currentSheet.value ?: return
        val transposedContent = ChordSheetTranspose.transpose(sheet.content, semitones)
        val updated = sheet.copy(
            content = transposedContent,
            updatedAt = System.currentTimeMillis(),
        )
        repository.save(updated)
        _currentSheet.value = updated
        refresh()
    }

    fun updateStrumPattern(patternName: String) {
        val sheet = _currentSheet.value ?: return
        val updated = sheet.copy(
            strumPatternName = patternName,
            updatedAt = System.currentTimeMillis(),
        )
        repository.save(updated)
        _currentSheet.value = updated
        refresh()
    }

    fun duplicateSheet(sheet: ChordSheet) {
        val now = System.currentTimeMillis()
        val copy = ChordSheet(
            title = "${sheet.title} (Copy)",
            subtitle = sheet.subtitle,
            artist = sheet.artist,
            content = sheet.content,
            key = sheet.key,
            capo = sheet.capo,
            strumPatternName = sheet.strumPatternName,
            labels = sheet.labels,
            createdAt = now,
            updatedAt = now,
        )
        repository.save(copy)
        _currentSheet.value = copy
        _isEditing.value = false
        refresh()
    }

    private val _selectedIds = MutableStateFlow<Set<String>>(emptySet())
    val selectedIds: StateFlow<Set<String>> = _selectedIds.asStateFlow()

    private val _isSelectionMode = MutableStateFlow(false)
    val isSelectionMode: StateFlow<Boolean> = _isSelectionMode.asStateFlow()

    fun toggleSelection(id: String) {
        val current = _selectedIds.value
        _selectedIds.value = if (id in current) current - id else current + id
        if (_selectedIds.value.isEmpty()) _isSelectionMode.value = false
    }

    fun selectAll() {
        _selectedIds.value = _sheets.value.map { it.id }.toSet()
    }

    fun clearSelection() {
        _selectedIds.value = emptySet()
        _isSelectionMode.value = false
    }

    fun enterSelectionMode(id: String) {
        _isSelectionMode.value = true
        _selectedIds.value = setOf(id)
    }

    fun deleteSelected() {
        _selectedIds.value.forEach { repository.delete(it) }
        _selectedIds.value = emptySet()
        _isSelectionMode.value = false
        refresh()
    }

    fun applyLabelsToSelected(labels: List<String>) {
        val ids = _selectedIds.value
        _allSheets.value.filter { it.id in ids }.forEach { sheet ->
            val merged = (sheet.labels + labels).distinct()
            val updated = sheet.copy(labels = merged, updatedAt = System.currentTimeMillis())
            repository.save(updated)
        }
        _selectedIds.value = emptySet()
        _isSelectionMode.value = false
        refresh()
    }

    fun deleteSheet(id: String) {
        repository.delete(id)
        if (_currentSheet.value?.id == id) {
            _currentSheet.value = null
        }
        refresh()
    }

    /**
     * Imports a song from ChordPro-formatted text.
     *
     * Parses the content, saves the resulting [ChordSheet], and opens it.
     *
     * @param content The raw ChordPro text.
     * @param filename Optional filename used as a fallback title.
     */
    fun importChordPro(content: String, filename: String? = null) {
        val sheet = ChordProParser.parse(content, titleFromFilename(filename))
        repository.save(sheet)
        _currentSheet.value = sheet
        _isEditing.value = false
        refresh()
    }

    /**
     * Imports a song from plain text (non-ChordPro).
     *
     * Wraps the content in a new [ChordSheet] and saves it.
     *
     * @param content The raw text content (may contain [ChordName] markers).
     * @param filename Optional filename used as the title.
     */
    fun importPlainText(content: String, filename: String? = null) {
        val sheet = ChordSheet(title = titleFromFilename(filename), content = content.trim())
        repository.save(sheet)
        _currentSheet.value = sheet
        _isEditing.value = false
        refresh()
    }

    /**
     * Derives a song title from an imported file name.
     *
     * Strips any path and Storage Access Framework document-ID prefix (e.g.
     * `primary:Download/`) before dropping the extension, so a URI or document ID
     * can never surface as the title (issue #500).
     *
     * @param filename The resolved file name, or `null` when unavailable.
     * @return A display title, falling back to [DEFAULT_IMPORT_TITLE].
     */
    private fun titleFromFilename(filename: String?): String =
        filename
            ?.substringAfterLast('/')
            ?.substringAfterLast(':')
            ?.substringBeforeLast('.')
            ?.replace('_', ' ')
            ?.trim()
            ?.takeIf { it.isNotEmpty() }
            ?: DEFAULT_IMPORT_TITLE

    private companion object {
        const val DEFAULT_IMPORT_TITLE = "Imported Song"
    }
}
